src/fit.jl

```diff
@@ -433,28 +416,241 @@ function ews_to_abundances(atm, linelist, A_X, measured_EWs, ew_window_size::Rea
+   listed above. The solver stops when all residuals are less than the corresponding tolerance.
+- `max_step_sizes` (default: `[1000.0, 1.0, 0.3, 0.5]`) is the maximum step size to take in each 
+   parameter direction.  This is used to prevent the solver from taking too large of a step and 
+   missing the solution.  Be particularly cautious with the vmic (third) parameter, as the 
+   unadjusted step size is often too large.
+- `parameter_ranges` (default: `[(2800.0, 8000.0), (-0.5, 5.5), (1e-3, 10.0), (-2.5, 1.0)]`) is the 
+   allowed range for each parameter. This is used to prevent the solver from wandering into 
+   unphysical parameter space, or outside the range of the MARCS grid supported by 
+   [`Korg.interpolate_marcs`](@ref). Note that vmic must be nonzero in order to avoid null derivatives.
+- `callback`:  is a function which is called at each step of the optimization.  
+   It is passed three arguments: 
+    - the current values of the parameters
+    - the residuals of each equation being solved
+    - the abundances of each line computed with the current parameters.
+  You can pass a callback function, to e.g. make a plot of the residuals at each step. 
+
+"""
+function ews_to_stellar_parameters(linelist, measured_EWs, measured_EW_err=ones(length(measured_EWs)); 
+                                   Teff0=5000.0, logg0=3.5, vmic0=1.0, metallicity0=0.0,
+                                   tolerances=[1e-3, 1e-3, 1e-4, 1e-3],
+                                   max_step_sizes=[1000.0, 1.0, 0.3, 0.5],
+                                   parameter_ranges=[(2800.0, 8000.0), (-0.5, 5.5), (1e-3, 10.0), (-2.5, 1.0)],
+                                   fix_params=[false, false, false, false],
+                                   callback=Returns(nothing), passed_kwargs...)
+    if length(linelist) != length(measured_EWs)
+        throw(ArgumentError("length of linelist does not match length of ews ($(length(linelist)) != $(length(measured_EWs)))"))
+    end
+    formulas = [line.species.formula for line in linelist]
+    if any(Ref(formulas[1]) .!= formulas)
+        throw(ArgumentError("All lines must be from the same element."))
+    end
+    if Korg.ismolecule(linelist[1].species)
+        throw(ArgumentError("Cannot do stellar parameter determination with molecular lines."))
+    end
+    neutrals = [l.species.charge == 0 for l in linelist]
+    if (sum(neutrals) < 3) || (sum(.! neutrals) < 1)
+        throw(ArgumentError("Must have at least 3 neutral lines and 1 ion line."))
+    end
+    if vmic0 == 0.0
+        throw(ArgumentError("Starting guess for vmic (vmic0) must be nonzero."))
+    end
+
+    # set up closure to compute residuals
+    get_residuals = (params) -> 
+        _stellar_param_equation_residuals(params, linelist, measured_EWs, measured_EW_err, 
+                                           fix_params, callback, passed_kwargs)
+
+    params = [Teff0, logg0, vmic0, metallicity0]
```

Comment:
I it sensible to clamp the initial parameters to be within `parameter_ranges` and show a warning if the initial value is outside the range? At the moment the behaviour is that an exception will be raised when trying to create an atmosphere. Maybe that's OK, but since there are explicit bounds for parameters, I would think it's sensible to clamp the initial parameters to be within the parameter bounds, and optionally show a warning.